### PR TITLE
Add support for basic named params in partials

### LINKED ISSUE #165: Allow simple named parameters in partials
E.g., a partial would include a template variable like `{{ $var }}` and you would include a partial with the syntax `(!/docs/pages/includes/myinclude.mdx var=value!)`. For simplicity, values would need to be strings.

The use case would be blocks of text we could reuse in the docs except for a single value. One example that comes to mind is explaining how to assign a Teleport role to a user when we need to mention a specific role name: the instructions are the same across docs guides, but the role name we use in the instructions varies from guide to guide.

### PR DESCRIPTION
**Pull request by ptgott**
Closes #165

Allow a partial inclusion expression, e.g., `(!mypartial.mdx!)` to assign named parameters that replace template variables in partials. For example, `(!mypartial.mdx var="value"!)` replaces the template variable `{{ var }}` in `mypartial.mdx` with `value`.

Design details:

- Variable values must be in quotes, single or double, and the quotes must match.

- There can be multiple variables/values. The inclusion expression separates these with spaces, e.g.:

  ``` (!mypartial.mdx var1="val" var2='val'!) ```

- Allows for default parameter values, which will make it easier to add parameter support to existing partials. E.g., with a template variable like this:

  ``` {{ myvar="myval" }} ```

  This partial will populate the template variable with `myval` at this particular location if the inclusion expression does not assign it a value.

- If a parameter does _not_ have a default value, and an inclusion expression does not provide one, the `{{ myvar }}` expression will be left untouched, allowing for situations where a partial needs to include a literal `{{ }}` string.

- Escapes: Unfortunately, due to some JavaScript quirks I didn't give myself time to look into, quotes must be escaped twice, e.g., `\\"`. Double escapes are removed within double-quoted strings but not within single-quoted strings.

### CODE REVIEW

**Comment by ptgott** [comment]
There's a proof of concept for using partial params in this branch in `gravitational/teleport`: `paul.gottschling/2022-10-add-role-partial`.

I need to resolve some linter issues, but thought I would open this first to get feedback on more substantial issues first.

**Comment by hatched** [comment]
What do you think about:
- making `(!mypartial.mdx var1="val" var2="val"!)` whitespace significant.
- only allowing a single quote type.
- using `(!foo="bar"!)` to set the default instead of `{{`

Making these changes should simplify the parser and also allow you to support multiple defaults `(!foo="bar" bax="baz"!)` with the same parser. I also feel it makes the syntax a bit more consistent.

```
The operating systems we support include {{ systems="macOS, Windows Server, and Linux" }}
```

would instead become

```
(!systems="macos, WindowsServer, and Linux"!)
The operating systems we support include {{ systems }}
```

**Comment by ptgott** [comment]
@hatched Thanks again for giving this a look!

> What do you think about:
> - making `(!mypartial.mdx var1="val" var2="val"!)` whitespace significant.

The current behavior is to separate parameter assignments with spaces (e.g., using it to [detect the start](https://github.com/gravitational/docs/blob/b0420e0d8e4c7ef459ab492d01c15fa29b6aeb67/server/remark-includes.ts#L110-L115) of a parameter name). Was there another kind of significant whitespace you were thinking of?

> - only allowing a single quote type.

This makes a lot of sense!

> - using `(!foo="bar"!)` to set the default instead of `{{`

The `{{` that sets the default is in the partial itself, rather than the page that's including the partial. Just to double-check, would we be using the `(! !)` syntax in both the page that's including the partial and the partial itself?

One possible issue is when one partial needs to include another partial (e.g., in [`start-auth-proxy.mdx`](https://github.com/gravitational/teleport/blame/master/docs/pages/includes/database-access/start-auth-proxy.mdx#L55)), but we could write a test case to prevent the `(!var="val"!)` inside a partial form being parsed as including another partial.

> 
> Making these changes should simplify the parser and also allow you to support multiple defaults `(!foo="bar" bax="baz"!)` with the same parser. I also feel it makes the syntax a bit more consistent.

With the current approach, the defaults are assigned at each position where a partial substitutes a single variable. For example, for this partial...

```mdx
This is a partial.

Here, we're substituting a word with {{ other="another word" }}.

Here, we're substituting a word with {{ other="yet another word" }}.
```

If a page includes this partial without assigning a value to `other` (e.g., `(!mypartial.mdx!)`), the docs engine would render:

```mdx
This is a partial.

Here, we're substituting a word with another word.

Here, we're substituting a word with yet another word.
```

To have multiple variables with multiple defaults, you would just declare another `{{ myvar="mydefault" }}` in the partial.

The alternative could be to include separate syntax within a partial that only exists to assign defaults, e.g.:

```mdx
{{ other="Another word" }}

Here, we're substituting a word with {{ other }}.

Here, we're substituting a word with {{ other }}.
```

In this case, the parser would remove the first line from the partial and use the default value to substitute `{{ other }}`.

Is that what you meant in the example below?

> ```
> The operating systems we support include {{ systems="macOS, Windows Server, and Linux" }}
> ```
> 
> would instead become
> 
> ```
> (!systems="macos, WindowsServer, and Linux"!)
> The operating systems we support include {{ systems }}
> ```

**Comment by hatched** [comment]
> > * making `(!mypartial.mdx var1="val" var2="val"!)` whitespace significant.
> 
> The current behavior is to separate parameter assignments with spaces (e.g., using it to [detect the start](https://github.com/gravitational/docs/blob/b0420e0d8e4c7ef459ab492d01c15fa29b6aeb67/server/remark-includes.ts#L110-L115) of a parameter name). Was there another kind of significant whitespace you were thinking of?

Sorry what I meant was that if superfluous spaces were added to that statement it would be invalid ex) `(!mypartial.mdx var1="val" var2 = "val"!)`

> > * using `(!foo="bar"!)` to set the default instead of `{{`
> 
> The `{{` that sets the default is in the partial itself, rather than the page that's including the partial. Just to double-check, would we be using the `(! !)` syntax in both the page that's including the partial and the partial itself?
> 
> One possible issue is when one partial needs to include another partial (e.g., in [`start-auth-proxy.mdx`](https://github.com/gravitational/teleport/blame/master/docs/pages/includes/database-access/start-auth-proxy.mdx#L55)), but we could write a test case to prevent the `(!var="val"!)` inside a partial form being parsed as including another partial.

So my thought was that because the `(!foo="bar"!)` statement didn't include a partial name it would only apply to the file it was defined in. I'm not too stuck on this, I just thought that because it's using a similar syntax you wouldn't need a special handler for `{{` defining the default values.

> The alternative could be to include separate syntax within a partial that only exists to assign defaults, e.g.:
> 
> ```gfm
> {{ other="Another word" }}
> 
> Here, we're substituting a word with {{ other }}.
> 
> Here, we're substituting a word with {{ other }}.
> ```
> 
> In this case, the parser would remove the first line from the partial and use the default value to substitute `{{ other }}`.

That's correct. I couldn't see a use case where `other` would (should?) have two different values within the same partial so having the ability to define it as a variable that can be overridden externally seemed like a good idea. ex)

```
(!systems="macos, WindowsServer, and Linux"!)

The operating systems we support include {{ systems }}
...
Remember that we only support {{ systems }} so if you're trying this on something else it'll fail.
```

would become:

```
The operating systems we support include macos, WindowsServer, and Linux
Remember that we only support macos, WindowsServer, and Linux so if you're trying this on something else it'll fail.
```

**Comment by ptgott** [comment]
> Sorry what I meant was that if superfluous spaces were added to that statement it would be invalid ex) (!mypartial.mdx var1="val" var2 = "val"!)

Sounds good—I've added this logic.

> That's correct. I couldn't see a use case where other would (should?) have two different values within the same partial so having the ability to define it as a variable that can be overridden externally seemed like a good idea. ex)

That also sounds good. I'll get this done next time I return to this.

**Review by avatus — APPROVED** [review]

File: server/remark-includes.ts [hunk 1]
```diff
@@ -44,6 +43,56 @@ export interface ParameterAssignments {
...
+
+  let assignments: ParameterAssignments = {};
+
+  // Since every assignment must end with a double quote and a space, split
+  // assignments on spaces that follow a non-escaped double quote.
+  const rawAssignments = assignmentSection.split(
+    new RegExp('(?<=[^\\\\]") ', "g")
+  );
+
+  const assignmentRegExp = new RegExp('([^ "]+)=(".*")', "g");
```
Comment (avatus) [inline server/remark-includes.ts:76]: nit: regex will be the bane of me. If we have a complicated-ish regex such as this, can we always leave a comment above it trying to explain it in plain english (as best as possible)

**Comment by avatus** [comment]
it looks great to me. Would wait to get the thumbs up from @hatched after the review feedback

**Review by hatched — DISMISSED** [review]
Looks good! Thanks for the updates.

**Comment by ptgott** [comment]
@hatched @avatus Thanks for the reviews! Just added a comment above the regexp—looks like my push dismissed the two approvals.

**Review by avatus — APPROVED** [review]

**Review by hatched — APPROVED** [review]

**Comment by ptgott** [comment]
Unfortunately, I'd forgotten that we needed to support NodeJS v14 for CI to pass, and `replaceAll` is only available beginning in v15. I've pushed another commit swapping `replaceAll` for `replace`.

File: server/remark-includes.ts [hunk 2]
```diff
@@ -197,8 +197,11 @@ export function resolveParamValue(val: string): string {
     );
   }
   let newVal = val.slice(1, val.length - 1);
+  const escapedQuoteRegexp = new RegExp('\\\\"', "g");
   if (val[0] == '"') {
-    newVal = newVal.replaceAll('\\"', '"');
+    newVal = newVal.replace(escapedQuoteRegexp, (escapedQuote) => {
```
Comment (avatus) [inline server/remark-includes.ts:202]: you shouldn't need to pass a function here if we are always returning the same value of `'"'`. Because your regex includes `g` in it, `replace` will work for all instances. I believe `newVal.replace(escapedQuoteRegexp, '"')` would work just fine. Same with the other. Is that not the case?

Comment (ptgott) [reply server/remark-includes.ts]: I misremembered the types of `replace` args that triggered a replacement of all instances and somehow thought you needed both a regexp and a function. I've changed this in 1cdd986569a44c52170be2950a9b4b1debb128ed.

**Review by hatched — APPROVED** [review]

File: server/remark-includes.ts (hunk 2 above)
Comment (avatus) [reply server/remark-includes.ts]: Although, to your original point: I did notice that vercel has our Docs repo at node 14. I wanted to switch to 16 but have been a bit scared to push the button. Let's talk about doing this sooner than later. FWIW, I've been running docs locally on 16 and no hiccups but you never know.

Comment (ptgott) [reply server/remark-includes.ts]: That sounds good to me! I've opened an issue here: https://github.com/gravitational/docs/issues/181
